<a href="https://colab.research.google.com/github/grzegorzbury/Wprowadzenie-do-Machine-Learningu-z-Pythonem-Tasks/blob/main/Chapter_1_Introduction_to_ML/1_6_Zadanie.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [41]:
from sklearn.datasets import load_wine
from sklearn.datasets import load_diabetes

from sklearn.tree import DecisionTreeRegressor
from sklearn.tree import DecisionTreeClassifier
from sklearn.ensemble import RandomForestClassifier
from sklearn.ensemble import RandomForestRegressor
from sklearn.linear_model import LinearRegression


from sklearn.model_selection import train_test_split

### Tworzenie klasyfikatorów

clf_tree_reg = DecisionTreeRegressor()
clf_tree_cls = DecisionTreeClassifier()
clf_rfor_cls = RandomForestClassifier()
clf_rfor_reg = RandomForestRegressor()
clf_line_reg = LinearRegression()

### Definicja funckji wczytywania danych z datesets

def import_wine_data():
    """
        Zwraca:
            X - dane
            y - target labels
            target_names - labels do tego zbioru
    """
    wine_data = load_wine()
    X, y = wine_data['data'], wine_data['target']
    return X, y

def import_V():
    """
        Zwraca:
            X - dane
            y - target labels
            target_names - labels do tego zbioru
    """
    diabetes = load_diabetes()
    X, y = diabetes['data'], diabetes['target']
    return X, y

Xw, yw = import_wine_data()
Xd, yd = import_V()

Xw_train, Xw_test, yw_train, yw_test = train_test_split(Xw, yw, test_size=0.20, random_state=0)
Xd_train, Xd_test, yd_train, yd_test = train_test_split(Xd, yd, test_size=0.20, random_state=0)


clf_list = [clf_tree_reg, clf_tree_cls, clf_rfor_cls, clf_line_reg,clf_rfor_reg]

# Test dla wine_dataset
for clf in clf_list:
    clf.fit(Xw_train, yw_train)
    
    yw_pred = clf.predict(Xw_test)

 #   print("true values ", str(clf), yw[:10])
 #   print("predicted   ", str(clf), yw_pred[:10])
    clf_score_train = clf.score(Xw_train, yw_train)
    print("[WINE] Score for the TRAIN: ", str(clf), clf_score_train)

    clf_score_test = clf.score(Xw_test, yw_test)
    print("[WINE] Score for the TEST: ", str(clf), clf_score_test)

# Test dla digits_dataset
for clf in clf_list:
    clf.fit(Xd_train, yd_train)
    
    yd_pred = clf.predict(Xd_test)

    print("true values ", str(clf), yd[:10])
    print("predicted   ", str(clf), yd_pred[:10])

    clf_score_train = clf.score(Xd_train, yd_train)
    print("[DIABETES] Score for the TRAIN: ", str(clf), clf_score_train)

    clf_score_test = clf.score(Xd_test, yd_test)
    print("[DIABETES] Score for the TEST: ", str(clf), clf_score_test)




[WINE] Score for the TRAIN:  DecisionTreeRegressor() 1.0
[WINE] Score for the TEST:  DecisionTreeRegressor() 0.8353658536585366
[WINE] Score for the TRAIN:  DecisionTreeClassifier() 1.0
[WINE] Score for the TEST:  DecisionTreeClassifier() 0.9722222222222222
[WINE] Score for the TRAIN:  RandomForestClassifier() 1.0
[WINE] Score for the TEST:  RandomForestClassifier() 0.9722222222222222
[WINE] Score for the TRAIN:  LinearRegression() 0.919071553176753
[WINE] Score for the TEST:  LinearRegression() 0.7643759153956304
[WINE] Score for the TRAIN:  RandomForestRegressor() 0.9929051761846902
[WINE] Score for the TEST:  RandomForestRegressor() 0.899644512195122
true values  DecisionTreeRegressor() [151.  75. 141. 206. 135.  97. 138.  63. 110. 310.]
predicted    DecisionTreeRegressor() [257. 306. 202. 202.  52. 277.  90. 212. 110. 220.]
[DIABETES] Score for the TRAIN:  DecisionTreeRegressor() 1.0
[DIABETES] Score for the TEST:  DecisionTreeRegressor() -0.4140435926556274
true values  DecisionTr

Do zadania załadowano dwa zbiory danych (1) wina (2) diabetycy. Zbiór wina to zbiór danych do klasyfikacji natomiast zbiór diabetyków odnosi się do przewidywania wartości (regresja). Wybrano 5 klasyfikatorów. Trzy z nich odnoszą się do regresji (DecisionTreeRegressor, RandomForestRegressor, LinearRegression) a dwa do klasyfikacji (RandomForestClassifier, DecisionTreeClassifier).

Taki wybór został dokonany w założeniu, że klasyfikatory zorientowane na regresję, przyniosą lepsze wyniki dla danych odnośnie do diabetyków, a klasyfikatory zorientowane na klasyfikację dadzą poprawne wyniki jedynie dla zbiorów opisującego wina.

Jednakże, założenia nie zostały całkowicie potwierdzone. Dla zbioru win wyróżnimy trzy klasyfikatory, dla których występuje overfitting dla danych uczących (DecisionTreeRegressor, DecisionTreeClassifier, RandomForestClassifier, ponieważ score = 1 co daje fałszywie wysoką skuteczność przewidywania dla danych treningowych (Testowano różny stosunek podziału; 0.20 oraz 0.33). Algorytm RandomForestRegressor również zbyt dobrze dopasowuje się do danych treningowych (score = 0.99). Podsumowując, dla zbioru win, najrozsądniejsze wyniki daje algorytm LinearRegression, ponieważ score dla treningowych i testowanych danych wynosi odpowiednio 0.92 i 0.76.

Dla danych o diabetykach wybrane modele nie są skuteczne. Po pierwsze algorytmy: DecisionTreeRegressor, DecisionTreeClassifier oraz RandomForestClassifier fałszywie dopasowują się do danych treningowych (score = 1) a następnie wykazują ekstremalnie niską jakość równą 0.0 lub 0.02. Algorytm DecisionTreeRegressor kompletnie nie nadaje się do opisu zbioru danych (nie da się wpasować danych w model), ponieważ score daje wynik negatywny. Algorytm LinearRegression daje wynik wyłącznie 0.55 dla danych treningowych, co owocuje w jakości danych testowych na poziomie 0.33. Rozsądny wynik dla danych uczących daje algorytm RandomForestRegressor (score = 0.92), jednak wynik dla danych testowych (0.29) pokazuje, że model nie jest skuteczny nawet w 30%.
